internal/ad/ad.go

```diff
@@ -491,116 +491,155 @@ func (ad *AD) parseGPOs(ctx context.Context, gpos []gpo, objectClass ObjectClass
 			Rules: make(map[string][]entry.Entry),
 		}
 		r = append(r, gpoWithRules)
-		if err := func() error {
-			ad.downloadables[name].mu.RLock()
-			defer ad.downloadables[name].mu.RUnlock()
-			_ = ad.downloadables[name].testConcurrent
-
-			log.Debugf(ctx, "Parsing GPO %q", name)
-
-			// We need to consider the uppercase version of the name as well,
-			// which could occur in some of the default GPOs such as Default
-			// Domain Policy.
-			classes := []string{"User", "USER"}
-			if objectClass == ComputerObject {
-				classes = []string{"Machine", "MACHINE"}
+		if err = ad.parseGPO(ctx, name, url, keyFilterPrefix, objectClass, gpoWithRules); err != nil {
+			return r, err
+		}
+	}
+
+	return r, nil
+}
+
+func (ad *AD) parseGPO(ctx context.Context, name, url, keyFilterPrefix string, objectClass ObjectClass, gpoWithRules policies.GPO) error {
+	ad.downloadables[name].mu.RLock()
+	defer ad.downloadables[name].mu.RUnlock()
+	_ = ad.downloadables[name].testConcurrent
+
+	log.Debugf(ctx, "Parsing GPO %q of class %q", name, objectClass)
+
+	// We need to consider the uppercase version of the name as well,
+	// which could occur in some of the default GPOs such as Default
+	// Domain Policy.
+	classes := []string{"User", "USER"}
+	if objectClass == ComputerObject {
+		classes = []string{"Machine", "MACHINE"}
+	}
+
+	var err error
+	var f *os.File
+	for _, class := range classes {
+		var e error
+		var files []os.DirEntry
+
+		policyDir := filepath.Join(ad.sysvolCacheDir, "Policies", filepath.Base(url), class)
+		files, e = os.ReadDir(policyDir)
+
+		if e != nil {
+			if errors.Is(e, fs.ErrNotExist) {
```

Comment:
I think we should can bypass those nested if / else by only one, and skip directly to the next class if we couldn’t read a directory.
However, if the error is different, it’s better to stop right away.

Something like what we had:
```go
if errors.Is(e, fs.ErrNotExist) {
    log.Debugf(ctx, "Policy directory %q not found", policyDir)
    continue
} else if e != nil {
  return e
}
```

I would drop then the next debug that we found something (we will already print this once try to look for a registry file).

Reply:
Suggestion applied.